smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php

```diff
@@ -0,0 +1,277 @@
+      );
+      return $ajax_response;
+    }
+
+    // Normalize subdomain.
+    // First, try to parse as full URL.
+    // If that fails, try to parse as subdomain.sendsmaily.net, and, then clean up subdomain and pass as is.
+    if (filter_var($subdomain, FILTER_VALIDATE_URL)) {
+      $url = parse_url($subdomain);
+      $parts = explode('.', $url['host']);
+      $subdomain = count($parts) >= 3 ? $parts[0] : '';
+    }
+    elseif (preg_match('/^[^\.]+\.sendsmaily\.net$/', $subdomain)) {
+      $parts = explode('.', $subdomain);
+      $subdomain = $parts[0];
+    }
+    $subdomain = preg_replace('/[^a-zA-Z0-9]+/', '', $subdomain);
+
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $subdomain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
+    );
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+    curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
+    $output = curl_exec($ch);
+    $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+    curl_close($ch);
+
+    switch ((int) $http_code) {
+      case 200:
+        // OK code, continue validateCredentials.
+        break;
+
+      case 401:
+        $ajax_response->addCommand(
+          new HtmlCommand(
+            '.smaily_message',
+            '<div class="messages messages--error">' . 'Credentials invalid.' . '</div>')
+        );
+        return $ajax_response;
+
+      case 404:
+        $ajax_response->addCommand(
+          new HtmlCommand(
+            '.smaily_message',
+            '<div class="messages messages--error">' . 'Subdomain error' . '</div>')
+        );
+        return $ajax_response;
+
+      default:
+        $ajax_response->addCommand(
+          new HtmlCommand(
+            '.smaily_message',
+            '<div class="messages messages--error">' . 'Something went wrong.' . '</div>')
+        );
+        return $ajax_response;
+    }
+
+    $autoresponders = json_decode($output, TRUE);
+    $autoresponder_list = [];
+    if (!empty($autoresponders)) {
+      foreach ($autoresponders as $autoresponder) {
+        if (!empty($autoresponder['id']) && !empty($autoresponder['title'])) {
+          $autoresponder_list[$autoresponder['id']] = trim($autoresponder['title']);
+        }
+      }
+    }
+
+    $form['smaily_container']['smaily_autoresponder']['#options'] = $autoresponder_list;
+    // Refreshes $form to display autoresponders.
+    $ajax_response->addCommand(new ReplaceCommand(NULL, $form));
+
+    $this->config('smaily_for_drupal.adminsettings')
+      ->set('smaily_domain', $subdomain)
+      ->set('smaily_username', $username)
+      ->set('smaily_password', $password)
+      ->save();
+    $ajax_response->addCommand(
+      new HtmlCommand(
+        '.smaily_message',
+        '<div class="messages messages--status">' . 'Credentials valid' . '</div>')
+    );
+    return $ajax_response;
+
+  }
+
```

Comment:
Is this whitespace a Drupal coding standard?

Reply:
Line 276 yes.

Reply:
OK.